nimare/meta/models.py

```diff
@@ -0,0 +1,1229 @@
+import functorch
+import numpy as np
+import pandas as pd
+import torch
+
+LGR = logging.getLogger(__name__)
+
+
+class GeneralLinearModelEstimator(torch.nn.Module):
+    """Base class for GLM estimators.
+
+    Parameters
+    ----------
+    spatial_coef_dim : :obj:`int`
+        Number of spatial B-spline bases. Default is None.
+    moderators_coef_dim : :obj:`int`, optional
+        Number of study-level moderators. Default is None.
+    penalty : :obj:`bool`
+        Whether to Firth-type regularization term. Default is False.
+    lr : :obj:`float`
+        Learning rate. Default is 0.1.
+    lr_decay : :obj:`float`
+        Learning rate decay for each iteration. Default is 0.999.
+    n_iter : :obj:`int`
+        Maximum number of iterations. Default is 1000.
+    tol : :obj:`float`
+        Tolerance for convergence. Default is 1e-2.
+    device : :obj:`str`
+        Device to use for computations. Default is "cpu".
+    """
+
+    _hessian_kwargs = {
+        "create_graph": False,
+        "vectorize": True,
+        "outer_jacobian_strategy": "forward-mode",
+    }
+
+    def __init__(
+        self,
+        spatial_coef_dim=None,
+        moderators_coef_dim=None,
+        penalty=False,
+        lr=0.1,
+        lr_decay=0.999,
+        n_iter=1000,
+        tol=1e-2,
+        device="cpu",
+    ):
+        super().__init__()
+        self.spatial_coef_dim = spatial_coef_dim
+        self.moderators_coef_dim = moderators_coef_dim
+        self.penalty = penalty
+        self.lr = lr
+        self.lr_decay = lr_decay
+        self.n_iter = n_iter
+        self.tol = tol
+        self.device = device
+
+        # initialization for iteration set up
+        self.iter = 0
+
+        # after fitting, the following attributes will be created
+        self.spatial_regression_coef = None
+        self.spatial_intensity_estimation = None
+        self.moderators_coef = None
+        self.moderators_effect = None
+        self.spatial_regression_coef_se = None
+        self.log_spatial_intensity_se = None
+        self.spatial_intensity_se = None
+        self.se_moderators = None
+
+    @abc.abstractmethod
+    def _log_likelihood_single_group(self, **kwargs):
+        """Log-likelihood of a single group.
+
+        Returns
+        -------
+        torch.Tensor
+            Value of the log-likelihood of a single group.
+        """
+        return
+
+    @abc.abstractmethod
+    def _log_likelihood_mult_group(self, **kwargs):
+        """Total log-likelihood of all groups in the dataset.
+
+        Returns
+        -------
+        torch.Tensor
+            Value of total log-likelihood of all groups in the dataset.
+        """
+        return
+
+    @abc.abstractmethod
+    def forward(self, **kwargs):
+        """Define the loss function (nagetive log-likelihood function) for each model.
+
+        Returns
+        -------
+        torch.Tensor
+            Value of the log-likelihood of a single group.
+        """
+        return
```

Comment:
the other abstract methods in NiMARE use `pass` instead of `return`, no functional difference, but using `pass` is more consistent with the rest of the codebase

Reply:
I've changed it to `return`